### DISCUSSION TITLE
How can I display the results of BookmarkShowAll from MattesGroeger/vim-bookmarks within fzflua?

### DISCUSSION THREAD

**Discussion by gengdz** [ibhagwan/fzf-lua#1275] [Q&A] [answered]
Previously, I utilized https://github.com/MattesGroeger/vim-bookmarks in conjunction with Telescope for viewing all my bookmarks. 

Now, having switched to fzflua, I am looking for a way to display the results of the BookmarkShowAll command within fzflua.

what should i do

**Comment by ibhagwan** [comment, marked as answer]
I’m not familiar with the plugin mentioned but the answer is the same, find a way to list all the bookmarks (plugin API if such exists or text parsing) and use fzf_exec to feed these into fzf.

Go to the Wiki and read the advanced API section and you’ll understand how to use fzf_exec, it’s very simple:
https://github.com/ibhagwan/fzf-lua/wiki/Advanced

**Reply by gengdz** [reply]
Thank you very much, I'll go give it a try.

**Reply by gengdz** [reply]
If the `lines` contain only the filename, then `fzf_lua.actions.file_edit` works. 

However, if I want to display it as padded_text | filename, then `fzf_lua.actions.file_edit` will not work.

only filename

![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/6e971e81-a197-4936-bcf4-ce2c5b5d712d)

padded_text |  filename

![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/0e6e485b-c5d5-431d-8a97-293cd9a78683)

```lua
      {
        "ma",
        function()
          local fzf_lua = require("fzf-lua")
          local bookmarks = get_bookmarks(vim.fn["bm#all_files"](), {})

          local lines = {}
          for _, bookmark in ipairs(bookmarks) do
            local padded_text = pad_right(bookmark.text, 60)
            -- local line = string.format("%s | %s", padded_text, bookmark.filename)
            local line = string.format("%s | %s", padded_text, bookmark.filename)
            -- table.insert(lines, line)
            table.insert(lines, bookmark.filename)
          end

          fzf_lua.fzf_exec(lines, {
            prompt = "Bookmarks> ",
            actions = {
              ["default"] = fzf_lua.actions.file_edit,
              ["default-other-way-only-for-you-see-my-code"] = {
                function(selected)
                  fzf_lua.actions.file_edit({ cwd = "1 1 2 /Users/xingya/.config/.bashrc" })
                end,
              },
            },
          })
        end,
        desc = "Bookmarks all marks",
      },

```

What should I do to make it jump correctly instead of throwing an error?

**Reply by ibhagwan** [reply]
Parse the line using `string.xxx` commands and extract the file name and then either send it to the filed edit action as  `actions.file_edit({ file }, opts)` or call your own edit using `vim.cmd("e " .. filename)`.

**Reply by gengdz** [reply]
like this?

```lua
  ["default"] = {
    function(selected)
      fzf_lua.actions.file_edit({ file = "/Users/xingya/.config/nvim/lazyvim.json" })
    end,
```

I got this result, and it seems there's something wrong with it.
![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/bfe795f8-0293-4b96-b75d-c08703bf8cfc)

**Reply by ibhagwan** [reply]
Not like that, first extract the file from the string, then send it as the first index:
```lua

["default"] = {
    function(selected, opts)
      -- fix the regex as needed based on your text entry
      local file = selected[1]:match('| (.*)')
      fzf_lua.actions.file_edit({ file }, opts)
    end,
```

**Reply by gengdz** [reply]
Okay, I understand what you mean. I'm using a special case for a file here.

I found that it didn't open a new buffer but showed an interface like this. What I expect is to open this buffer, which is the default action for finding files.

![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/05629113-76aa-4ed0-90e2-c58f4f9aff4a)

```lua
    ["default"] = {
      function(selected, opts)
        -- fix the regex as needed based on your text entry
        -- local file = selected[1]:match("| (.*)")
        local file = "/Users/xingya/.config/nvim/lazyvim.json"
        fzf_lua.actions.file_edit({ file }, opts)
      end,
    },

```

**Reply by ibhagwan** [reply]
Because you wrapped the action in `{}` (which I also mistakenly copied from your code) implying a “resume” action, change it to this:
```lua

    ["default"] = function(selected, opts)
        -- fix the regex as needed based on your text entry
        -- local file = selected[1]:match("| (.*)")
        local file = "/Users/xingya/.config/nvim/lazyvim.json"
        fzf_lua.actions.file_edit({ file }, opts)
      end,
```

**Reply by gengdz** [reply]
Great, it works! I love you!

**Reply by ibhagwan** [reply]
Fantastic! Ty @gengdz  ❤️

**Reply by gengdz** [reply]
I can now jump to the file, so how do I jump to a specific line?

In the following code, `lnum` represents the specified line.
```lua
    actions = {
      ["default"] = function(selected, opts)
        -- fix the regex as needed based on your text entry
        -- local file = selected[1]:match("| (.*)")
        local lnum, text, file = selected[1]:match("(%d+)%s+|%s+(.-)%s+|%s+(.+)")
        fzf_lua.actions.file_edit({ file }, opts)
      end,
    },

```

**Reply by ibhagwan** [reply]
If you format the text the same as grep/rg output the action will do the jump for you, the format should be `PATH:LINE:COL`:
```lua
fzf_lua.actions.file_edit({ string.format('%s:%s:%s', file, line, col) }, opts)
```

**Reply by gengdz** [reply]
I'll try it now.

**Reply by gengdz** [reply]
It works, thank you very, very much.

**Reply by gengdz** [reply]
Another question： I want to add a preview feature, but now the display is incorrect.

```lua
  fzf_lua.fzf_exec(lines, {
    winopts = {
      title = "Bookmarks",
      title_pos = "center",
      preview = {
        horizontal = "right:40%",
        layout = "horizontal",
      },
    },
    previewer = {},
    actions = {
      ["default"] = function(selected, opts)
        local line, text, file = selected[1]:match("(%d+)%s+|%s+(.-)%s+|%s+(.+)")
        fzf_lua.actions.file_edit({ string.format("%s:%s:%s", file, line, 1) }, opts)
      end,
    },
  })

```
![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/1fd0cc39-2269-4c4f-aae4-df465a47e873)

**Reply by ibhagwan** [reply]
You’ll need to create a custom previewer derived from the builtin previewer, and implement your own `parse_entry` method to supply the previewer class with what object to preview, see:
https://github.com/ibhagwan/fzf-lua/wiki/Advanced#preview-nvim-builtin

**Reply by gengdz** [reply]
I have completed this task, thank you so much for your help @ibhagwan 

![image](https://github.com/ibhagwan/fzf-lua/assets/41912568/b4032722-00f6-4aed-8871-3e1811dd1c15)

code: https://github.com/gengdz/dotfiles/blob/741c94940ecf3d9c0f7a37fd33cf1d72f09c75b7/nvim/lua/plugins/editor.lua#L106

**Reply by ibhagwan** [reply]
Looks beautiful @gengdz!
